src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java

```diff
@@ -251,57 +262,95 @@ private FormData prepareFormDataSet(FormFields form) {
+		fieldAttributes.putAll(field.getFormControl().getAttributesMap());
+
+		// Places a key, Control Type, for each FormControlType
+		fieldAttributes.put("Control Type", field.getFormControl().getFormControlType().name());
+
+		//If the field has a value attribute present(Predefined value)
+		//Should store the value being submitted to be passed to the ValueGenerator
+		if(field.getFormControl().getAttributesMap().containsKey("value")){
+			defaultValue = field.getFormControl().getAttributesMap().get("value");
+		}
+
+		//Handles Submit Fields
+		if (field.getFormControl().getFormControlType().isSubmit()) {
+			List<String> submitFields = new ArrayList<String>();
+			for (String value : field.getPredefinedValues()){
+				finalValue = this.valueGenerator.getValue(uri, url, fieldId, value, definedValues, envAttributes, fieldAttributes);
+				submitFields.add(finalValue);
+			}
+			return submitFields;
+		}
+
 		if (log.isDebugEnabled()) {
 			log.debug("Existing values: " + values);
 		}
 
 		// If there are no values at all or only an empty value
 		if (values.isEmpty() || (values.size() == 1 && values.get(0).isEmpty())) {
-			String finalValue = DEFAULT_EMPTY_VALUE;
+			finalValue = DEFAULT_EMPTY_VALUE;
+			defaultValue = finalValue;
 
 			// Check if we can use predefined values
 			Collection<String> predefValues = field.getPredefinedValues();
 			if (!predefValues.isEmpty()) {
+				//Store those predefined values in a list for the DefaultValueGenerator
+				definedValues.addAll(predefValues);
 				// Try first elements
 				Iterator<String> iterator = predefValues.iterator();
 				finalValue = iterator.next();
+				defaultValue = finalValue;
 
 				// If there are more values, don't use the first, as it usually is a "No select"
 				// item
 				if (iterator.hasNext()) {
 					finalValue = iterator.next();
-				}
-			} else {
-				/*
-				 * In all cases, according to Jericho documentation, the only left option is for
-				 * it to be a TEXT field, without any predefined value. We check if it has only
-				 * one userValueCount, and, if so, fill it with a default value.
-				 */
-				if (field.getUserValueCount() > 0) {
-					finalValue = getDefaultTextValue(field);
+					defaultValue = finalValue;
 				}
 			}
 
-			log.debug("No existing value for field " + field.getName() + ". Generated: " + finalValue);
-
-			values = new ArrayList<>(1);
-			values.add(finalValue);
+		} else {
+			defaultValue = values.get(0);
 		}
 
+		/*
+		 * In all cases, according to Jericho documentation, the only left option is for
+		 * it to be a TEXT field, without any predefined value. We check if it has only
+		 * one userValueCount, and, if so, fill it with a default value.
+		 */
+
+		//Get the default value used in DefaultValueGenerator
+		finalValue = this.valueGenerator.getValue(uri, url, fieldId, defaultValue, definedValues, envAttributes, fieldAttributes);
+
+		//Reset the Default Value to empty for each field
+		defaultValue = "";
```

Comment:
This could be removed, it's no longer needed (it's not used after calling `getValue` above).